module/zfs/sa.c

```diff
@@ -451,6 +451,54 @@ sa_add_layout_entry(objset_t *os, sa_attr_type_t *attrs, int attr_count,
 	return (tb);
 }
 
+static sa_lot_t *
+sa_lookup_cached_layout(sa_os_t *sa, uint64_t hash, sa_attr_type_t *attrs,
+    int count)
+{
+	sa_layout_cache_t *slc;
+	sa_lot_t *tb = NULL;
+	int i, j;
+
+	kpreempt_disable();
+	slc = sa->sa_layout_cache + CPU_SEQID;
+
+	for (i = 0; i < SA_LAYOUT_ENTRIES; i++) {
+		if (slc->slc_lot[i] == NULL)
+			break;
+		if (slc->slc_lot[i]->lot_hash != hash)
+			continue;
+		if (sa_layout_equal(slc->slc_lot[i], attrs, count) == 0) {
+			tb = slc->slc_lot[i];
+			break;
+		}
+	}
+	if (tb != NULL) {
+		/* shift toward the most recently used */
+		j = (i + 1) % SA_LAYOUT_ENTRIES;
+		if (j != slc->slc_head) {
+			sa_lot_t *t = slc->slc_lot[i];
+			ASSERT(j < SA_LAYOUT_ENTRIES);
+			slc->slc_lot[i] = slc->slc_lot[j];
+			slc->slc_lot[j] = t;
+		}
+	}
+	kpreempt_enable();
+
+	return (tb);
+}
+
+static void
+sa_add_cached_layout(sa_os_t *sa, sa_lot_t *tb)
+{
+	sa_layout_cache_t *slc;
+
+	kpreempt_disable();
+	slc = sa->sa_layout_cache + CPU_SEQID;
+	slc->slc_lot[slc->slc_head] = tb;
+	slc->slc_head = (slc->slc_head + 1) % SA_LAYOUT_ENTRIES;
+	kpreempt_enable();
+}
+
 static void
 sa_find_layout(objset_t *os, uint64_t hash, sa_attr_type_t *attrs,
```

Comment:
the majority of Lustre use cases are HPC applications when many nodes share very similar (if not same) naming rules, stripes, etc. so when a cluster runs such an application MDS should be serving files with very few different EAs (because striping and naming schema are the same), so the most of requests should hit LAYOUT/IDX cache. in some cases there may be few big applications or lots of small ones when a cluster is shared among many groups. in this case I'd expect less hits, but still - most of applications work with MDS when an application is starting (open/create files) and is stopping (closing files, update attributes). these are relatively short periods and I think different applications don't overlap much and this should result in more cache hits.
answering your question - the more applications working with different striping and naming schemas, the less hits.
it'd very interesting to learn how many layouts/indices are used on real big sites. I guess it should be easy to count the numbers and dump via /proc. I'm fine to cook a patch. @behlendorf would it be possible to check on your production systems?